src/queue/multilevel.rs

```diff
@@ -0,0 +1,838 @@
+                let start_index = self.rng.gen_range(0, len);
+                for stealer in self
+                    .stealers
+                    .iter()
+                    .chain(&self.stealers)
+                    .skip(start_index)
+                    .take(len)
+                {
+                    match stealer.steal_batch_and_pop(&self.local_queue) {
+                        Steal::Success(t) => return Some(into_pop(t, false)),
+                        Steal::Retry => need_retry = true,
+                        _ => {}
+                    }
+                }
+            }
+            for injector in self
+                .level_injectors
+                .iter()
+                .chain(&*self.level_injectors)
+                .skip(expected_level + 1)
+                .take(LEVEL_NUM - 1)
+            {
+                match injector.steal_batch_and_pop(&self.local_queue) {
+                    Steal::Success(t) => return Some(into_pop(t, false)),
+                    Steal::Retry => need_retry = true,
+                    _ => {}
+                }
+            }
+            if !need_retry {
+                break None;
+            }
+        }
+    }
+}
+
+/// The runner builder for multilevel task queues.
+///
+/// It can be created by [`Builder::runner_builder`].
+pub struct MultilevelRunnerBuilder<B> {
+    inner: B,
+    manager: Arc<LevelManager>,
+}
+
+impl<B, R, T> RunnerBuilder for MultilevelRunnerBuilder<B>
+where
+    B: RunnerBuilder<Runner = R>,
+    R: Runner<TaskCell = T>,
+    T: TaskCell,
+{
+    type Runner = MultilevelRunner<R>;
+
+    fn build(&mut self) -> Self::Runner {
+        MultilevelRunner {
+            inner: self.inner.build(),
+            manager: self.manager.clone(),
+        }
+    }
+}
+
+/// The runner for multilevel task queues.
+///
+/// The runner helps multilevel task queues collect additional information.
+/// [`MultilevelRunnerBuilder`] is the [`RunnerBuilder`] for this runner.
+pub struct MultilevelRunner<R> {
+    inner: R,
+    manager: Arc<LevelManager>,
+}
+
+impl<R, T> Runner for MultilevelRunner<R>
+where
+    R: Runner<TaskCell = T>,
+    T: TaskCell,
+{
+    type TaskCell = T;
+
+    fn start(&mut self, local: &mut Local<T>) {
+        self.inner.start(local)
+    }
+
+    fn handle(&mut self, local: &mut Local<T>, mut task_cell: T) -> bool {
+        let extras = task_cell.mut_extras();
+        let running_time = extras.running_time.clone();
+        let level = extras.current_level;
+        let begin = Instant::now();
+        let res = self.inner.handle(local, task_cell);
+        let elapsed = begin.elapsed();
+        if let Some(running_time) = running_time {
+            running_time.inc_by(elapsed);
+        }
+        if level == 0 {
+            self.manager.level0_elapsed.inc_by(elapsed);
+        }
+        let current_total = Duration::from_micros(
+            self.manager.total_elapsed.inc_by(elapsed) + elapsed.as_micros() as u64,
```

Comment:
Why add twice?

Reply:
`fetch_add` returns the previous value.

Reply:
It's a bit confusing. I think we should make `inc_by` return the current value.